### PR TITLE
[OPPRO-347] Enable caching for Velox data connector

### PR DESCRIPTION
**Pull request by zhouyuan**
## What changes were proposed in this pull request?

This patch enables the caching feature for Velox. Cache file(s) will be generated on Spark context initialization, based on the configurations, velox will asynchronously caching the contents when reading from hdfs/s3.  The feature should be helpful if reading from slow/remote storage. 

Below configurations are added:
```
    spark.gluten.sql.columnar.backend.velox.cacheEnabled // enable or disable velox cache, default off
    spark.gluten.sql.columnar.backend.velox.cachePath  // the folder to store the cache files, default to /tmp
    spark.gluten.sql.columnar.backend.velox.cacheSize  // the total size of the cache, default to 1GB
    spark.gluten.sql.columnar.backend.velox.cacheShards // the shards of the cache, default to 1
    spark.gluten.sql.columnar.backend.velox.cacheIOThreads // the threads for cache promoting, default to 1

```
Note the current patch does not reuse the cache from last run - a new cache file will be generated for each new spark-context.

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

## How was this patch tested?

manually test

### CODE REVIEW

**Comment by zhouyuan** [comment]
Hi @akashsha1 could you please help to review this?

**Comment by akashsha1** [comment]
> Hi @akashsha1 could you please help to review this?

sure. @paul-amonson please review as well.

**Comment by paul-amonson** [comment]
I don't see any code issues but I do have a question:  The velox configuration variables, specifically "spark.gluten.sql.columnar.backend.velox.cacheEnabled", are not found in the facebook velox source tree or the special gluten velox source tree. How is configuration passed to and consumed by velox?

**Comment by paul-amonson** [comment]
> I don't see any code issues but I do have a question: The velox configuration variables, specifically "spark.gluten.sql.columnar.backend.velox.cacheEnabled", are not found in the facebook velox source tree or the special gluten velox source tree. How is configuration passed to and consumed by velox?

This was answered out of band. Thanks.

**Comment by FelixYBW** [comment]
Can we merge this PR?

**Review by akashsha1 — APPROVED** [review]
Looks good. signing off. thanks Yuan.

File: cpp/velox/compute/VeloxPlanConverter.cc [hunk 1]
```diff
@@ -71,7 +78,36 @@ void VeloxInitializer::Init(std::unordered_map<std::string, std::string> conf) {
...
+    uint64_t memoryBytes = 20L << 30;
+    options.capacity = memoryBytes;
+
+    auto allocator = std::make_shared<memory::MmapAllocator>(options);
+    mappedMemory_ = std::make_shared<cache::AsyncDataCache>(allocator, memoryBytes, std::move(ssd));
+
+    // register as default instance, will be used in parquet reader
+    memory::MappedMemory::setDefaultInstance(mappedMemory_.get());
+    VELOX_CHECK_NOT_NULL(dynamic_cast<cache::AsyncDataCache*>(mappedMemory_.get()));
+    LOG(INFO) << "STARTUP: Using AsyncDataCache";
```
[the remark this answers is not in the document]
Comment (zhouyuan) [reply cpp/velox/compute/VeloxPlanConverter.cc]: fixed in [7bbfd50](https://github.com/oap-project/gluten/pull/632/commits/7bbfd50046b1c71bbe426fcb362b8228a832c178)

File: cpp/velox/compute/VeloxPlanConverter.cc [hunk 2]
```diff
@@ -71,7 +78,36 @@ void VeloxInitializer::Init(std::unordered_map<std::string, std::string> conf) {
...
+      if (k == kVeloxCacheSize)
+        cacheSize = stol(v);
+      if (k == kVeloxCacheShards)
+        cacheShards = stoi(v);
+      if (k == kVeloxCachePath)
+        cachePathPrefix = v;
+    }
+    std::string cachePath = cachePathPrefix + "/cache." + genUuid();
+    cacheExecutor_ = std::make_unique<folly::IOThreadPoolExecutor>(cacheShards);
+    auto ssd = std::make_unique<cache::SsdCache>(cachePath, cacheSize, cacheShards, cacheExecutor_.get());
```
[the remark this answers is not in the document]
Comment (zhouyuan) [reply cpp/velox/compute/VeloxPlanConverter.cc]: fixed in [54cec2f](https://github.com/oap-project/gluten/pull/632/commits/54cec2f075747611f01987d80d1917766eb276b2)

**Review by zhejiangxiaomai — APPROVED** [review]

**Comment by FelixYBW** [comment]
should we set spark.gluten.sql.columnar.backend.velox.cacheIOThreads  to 5 as Orri suggested? @akashsha1 what's the test result?

**Comment by akashsha1** [comment]
> should we set spark.gluten.sql.columnar.backend.velox.cacheIOThreads to 5 as Orri suggested? @akashsha1 what's the test result?

Don't have the result yet. Let's merge the PR first, and we can iterate based on finding to update the default value. 
@zhouyuan - can you provide me a gluten jar to validate the efficiency of this change?

**Pull request merged**